be/src/storage/tablet_updates.cpp

```diff
@@ -5686,49 +5686,36 @@ void TabletUpdates::_reset_apply_status(const EditVersionInfo& version_info_appl
 }
 
 void TabletUpdates::rewrite_rs_meta() {
-    std::unordered_map<int64_t, RowsetSharedPtr> pending_rs;
-    std::vector<RowsetSharedPtr> published_rs;
-    {
-        std::lock_guard lg(_lock);
-        for (auto& [_, rs] : _rowsets) {
-            if (rs->rowset_meta()->skip_tablet_schema()) {
-                published_rs.emplace_back(rs);
-            }
-        }
-
-        for (auto& [version, rs] : _pending_commits) {
-            if (rs->rowset_meta()->skip_tablet_schema()) {
-                pending_rs[version] = rs;
-            }
-        }
-    }
-
+    std::lock_guard lg(_lock);
     for (auto& [version, rs] : _pending_commits) {
-        RowsetMetaPB meta_pb;
-        rs->rowset_meta()->get_full_meta_pb(&meta_pb);
-        Status st = TabletMetaManager::pending_rowset_commit(
-                _tablet.data_dir(), _tablet.tablet_id(), version, meta_pb,
-                RowsetMetaManager::get_rowset_meta_key(_tablet.tablet_uid(), rs->rowset_id()));
-        LOG_IF(FATAL, !st.ok()) << "fail to save pending rowset meta:" << st << ". tablet_id=" << _tablet.tablet_id()
-                                << ", rowset_id=" << rs->rowset_id();
-        rs->rowset_meta()->set_skip_tablet_schema(false);
+        if (rs->rowset_meta()->skip_tablet_schema()) {
+            rs->rowset_meta()->set_skip_tablet_schema(false);
+            RowsetMetaPB meta_pb;
+            rs->rowset_meta()->get_full_meta_pb(&meta_pb);
+            Status st = TabletMetaManager::pending_rowset_commit(
+                    _tablet.data_dir(), _tablet.tablet_id(), version, meta_pb,
+                    RowsetMetaManager::get_rowset_meta_key(_tablet.tablet_uid(), rs->rowset_id()));
+            LOG_IF(FATAL, !st.ok()) << "fail to save pending rowset meta:" << st
```

Comment:
It should be FATAL, we can return st.

Reply:
And I think we can use `WriteBatch` to batch write to disk.

Reply:
> And I think we can use `WriteBatch` to batch write to disk.

done.